Background bulk file import, refs #10137

sbreker opened the pull request:
Created new arFileImportJob class to create the CSV and XML import objects
and ingest the files on the server in the background.

arFileImportJob will catch any exceptions thrown during the CSV or XML import
process and display them on the job details window.

Simplified post import importSuccess page as much on there was no longer
relevant now that import occurs in the background.

If more than one import file is specified on the WebUI import page,
modules/object/actions/importAction.class.php will launch a new
arFileImportJob for each file to be imported.

Removed logic pretaining to zip files from importAction.class.php.

sbreker committed:
Background bulk file import, refs #10137

Created new arFileImportJob class to create the CSV and XML import objects
and ingest the files on the server in the background.

arFileImportJob will catch any exceptions thrown during the CSV or XML import
process and display them on the job details window.

Simplified post import importSuccess page as much on there was no longer
relevant now that import occurs in the background.

If more than one import file is specified on the WebUI import page,
modules/object/actions/importAction.class.php will launch a new
arFileImportJob for each file to be imported.

Removed logic pretaining to zip files from importAction.class.php.

MikeFE commented on apps/qubit/modules/object/actions/importAction.class.php:
```diff
@@ -51,103 +51,30 @@ public function execute($request)
-    // set indexing preference
-    if (isset($request->noindex))
+    $options = array('noIndex' => ($request->getParameter('noindex') == 'on') ? false : true,
+                     'doCsvTransform' => ($request->getParameter('doCsvTransform') == 'on') ? true : false,
```
can you just remove the part after ? since it'll evaluate as bool anyway? or is it a bit weird since you're doing the expression in the array definition like that? Same thing with above except changing == to !=

sbreker replied:
I wrote it this way for clarity, as the arFileImportJob depends on it being true/false, as opposed to 'on' or ''.  I don't think there is a technical reason preventing me from shortening it.  Thoughts?

MikeFE replied:
It's a very minor point, if you feel this way is clearer, that's cool :+1:

MikeFE commented on lib/job/arFileImportJob.class.php:
```diff
@@ -0,0 +1,90 @@
+          if ($parameters['doCsvTransform']) $this->info($this->i18n->__('Applying transformation to CSV file.'));
+          $importer->doCsvTransform = $parameters['doCsvTransform'];
+
+          if ($parameters['noIndex']) $this->info($this->i18n->__('Indexing imported records.'));
```
We as a rule always use curly braces even when it's not strictly necessary with the 1-liners like this.

sbreker replied:
👍

sbreker committed:
Background bulk file import, refs #10137

WebUI changes to support file import enhancements:

- modified object module importSelectAction.class.php controller to extend
  DefaultEditAction.
- removed object module controller importAction.php.
- created checkReposFilter.js to blank out repos dropdown if collection
  autocomplete is populated.
- created panels.less in arDominionPlugin to get panel functionality.
- modified object module importSelectSuccess.php template to
  incorporate interface changes.

 On branch dev/issue-10137-bulk-import-jobs
 Changes to be committed:
	deleted:    apps/qubit/modules/object/actions/importAction.class.php
	modified:   apps/qubit/modules/object/actions/importSelectAction.class.php
	modified:   apps/qubit/modules/object/config/view.yml
	renamed:    apps/qubit/modules/object/templates/importSuccess.php -> apps/qubit/modules/object/templates/importResultsSuccess.php
	modified:   apps/qubit/modules/object/templates/importSelectSuccess.php
	new file:   js/checkReposFilter.js
	new file:   plugins/arDominionPlugin/css/less/panels.less
	modified:   plugins/arDominionPlugin/css/main.less

sbreker committed:
Background bulk file import, refs #10137

Integrated WebUI changes to backend import logic. Updated
checkReposFilter.js to contain additional logic for controlling which fields
are available based on options selected.

- arFileImportJob.class.php now stages import params to pass into CSV and XML
import classes.
- QubitCsvImport.class.php modified to integrate new WebUI fields to the CLI
command string.
- Additional changes to importSelectSuccess.php and checkReposFilter.js to
capture logic surounding which fields should be present based on object type
being imported, and what other fields are selected.

 On branch dev/issue-10137-bulk-import-jobs
 Changes to be committed:
	modified:   apps/qubit/modules/object/actions/importSelectAction.class.php
	modified:   apps/qubit/modules/object/templates/importSelectSuccess.php
	modified:   js/checkReposFilter.js
	modified:   lib/QubitCsvImport.class.php
	modified:   lib/job/arFileImportJob.class.php

MikeFE commented on apps/qubit/modules/object/actions/importSelectAction.class.php:
```diff
@@ -31,29 +37,162 @@ public function execute($request)
+        $criteria->addAscendingOrderByColumn('authorized_form_of_name');
+        $cache = QubitCache::getInstance();
+        $cacheKey = 'file-import:list-of-repositories:'.$this->context->user->getCulture();
+        //$cacheKey = 'advanced-search:list-of-repositories:'.$this->context->user->getCulture();
```
this is no longer needed, right?

sbreker replied:
The commented out line?  :)   Nope - I will remove that.

MikeFE commented on apps/qubit/modules/object/actions/importSelectAction.class.php:
```diff
@@ -31,29 +37,162 @@ public function execute($request)
+
+    if ($request->isMethod('post'))
+    {
+      //$this->getParameters = $request->getGetParameters();
```
not needed?

MikeFE commented on apps/qubit/modules/object/actions/importSelectAction.class.php:
```diff
@@ -31,29 +37,162 @@ public function execute($request)
+        $this->setTemplate('importResults');
+      }
+    }
+
```
we usually don't use 2 newlines before starting elses

MikeFE commented on lib/QubitCsvImport.class.php:
```diff
@@ -80,42 +90,74 @@ public function import($csvFile, $type = null, $csvOrigFileName = null)
+        case 'delete-and-replace':
+          $commandUpdate = ($this->updateType == 'match-and-update') ? '--update="match-and-update"' : '--update="delete-and-replace"';
+          $commandSkipUnmatched = ($this->skipUnmatched) ? '--skip-unmatched' : '';
+          $commandLimit = ('' !== $this->limit) ? "--limit=\"$this->limit\"" : '';
```
`'--limit="$this->limit"'` ?

sbreker replied:
Changing to single quotes there will print the literal: **--limit="$this->limit"**
as opposed to evaluating $this->limit: **--limit="museum-of-agriculture"**

MikeFE replied:
fair enough

MikeFE commented on lib/job/arFileImportJob.class.php:
```diff
@@ -0,0 +1,190 @@
+    return true;
+  }
+
+  private function setCsvImportParams(&$importer, $parameters)
```
it's usually good for non-trivial functions to have a comment block at the top summarizing parameters/return values and the function's purpose. You can see examples elsewhere in the codebase on how we do it if you're not certain

MikeFE commented on lib/job/arFileImportJob.class.php:
```diff
@@ -0,0 +1,190 @@
+  {
+    foreach ($parameters as $key => $value)
+    {
+      if (empty($value)) { continue; }
```
curly braces should always be on their own line

MikeFE commented on lib/job/arFileImportJob.class.php:
```diff
@@ -0,0 +1,190 @@
+
+    foreach ($parameters as $key => $value)
+    {
+      if (empty($value)) { continue; }
```
same as above
